src/index.ts

```diff
@@ -1,87 +1,90 @@
-import { useState, useCallback, useEffect } from "react";
-import update from "immutability-helper";
+import {
+  useMemo,
+  useEffect,
+  useReducer,
+  useCallback,
+} from "react";
 
-import { HookArguments, HookReturnValues, Item } from "./types";
+import reducer from "./reducer";
+import {
+  HookArguments,
+  DefaultItem,
+  ActionType,
+  State,
+} from "./types";
+import { INITIAL_STATE } from "./constants";
 
-export const DEFAULT_ITEM_IDENTIFIER = "id";
-
-function useSelectedItems<T extends Record<any, any>>({
-  items = [],
-  itemIdentifier = DEFAULT_ITEM_IDENTIFIER,
-}: HookArguments<T>): HookReturnValues<T> {
-  const [selectedItems, setSelectedItems] = useState<T[]>([]);
-  const [itemsList, setItemsList] = useState<Item<T>[]>([]);
+function useSelectedItems<T extends DefaultItem, K>({
+  initialItems = [],
+  itemIdentifierKey,
+  initialSelectedItems = [],
+}: HookArguments<T, K>) {
+  const [items, dispatch] = useReducer(
+    reducer,
+    INITIAL_STATE,
+    (state: State) => ({
+      ...state,
+      itemIdentifierKey,
+    }),
+  );
 
   useEffect(() => {
-    const itemIdentifierIsValid = items.some((findItem: T) => (
-      findItem[itemIdentifier]
-    ));
+    const itemsWithInvalidIdentifers = initialItems.filter(
+      (findItem: T) => {
+        const hasItemIdentifierKey = Boolean(findItem[itemIdentifierKey]);
```

Comment:
Makes sense to use `Boolean` here? Values like `0`, `''` (empty string), and other things can't be used as `id`. Maybe `hasOwnProperty` or could check for null/undefined?

Reply:
Exactly, `0` and `''` can't be used as a valid `id`. 

```
const hasItemIdentifierKey = Boolean("") // false
``` 

I prefer to use ``Boolean`` to parse the value instead of using `!!`, it's more intuitive

Reply:
Wait, I think I didn't understand exactly. Do you mention that `''` and `0` could be used as ids, right?

Reply:
So yeah, that makes a lot of sense. I would rather check if it's ``null`` or ``undefined``

Reply:
I really didn't think we have a use case for an `empty string`, I said more like an example. For me makes sense the libraries by more agnostic about the way the user handler her id. Some examples: we can have an array called options and generate the items in memory using the index of the option to be the id (and this wouldn't have problems with React if the options are statically generated, they could ordering the options, etc, because it's a unique id in this context). Or they could have this array called options and the default option has id as null (because that I say `hasOwnProperty` could be a solution), because all the other options have id as number and to force the system to checks by id, you can't use just the id as 0.

Reply:
Improved on [`ec4fb9d` (#11)](https://github.com/LauraBeatris/use-selected-items-hook/pull/11/commits/ec4fb9d4bb3be159be7e4bb54c2ac8936166785a)